### PR TITLE
Dawn update

### PR DESCRIPTION
**Pull request by ambergorzynski**
Updates build and program generation to be compatible with a more recent version of Dawn (commit [f5d6487f00](https://dawn.googlesource.com/dawn/+/f5d6487f00dd64af4ef0bd846b578282296ef3c3)). Also adds a build option to specify a particular Dawn version via a filepath to a folder containing Dawn.

### CODE REVIEW

**Comment by hasali19** [comment]
Thanks for the PR! Please could you bump the submodule in `external/dawn` to whichever dawn commit you've tested this with.

File: crates/dawn/build.rs [hunk 1]
```diff
@@ -24,56 +25,35 @@ fn main() -> Result<(), Box<dyn Error>> {
...
                 .status()?
                 .success()
         {
             panic!("ar command failed");
         }
 
-        println!("cargo:rerun-if-changed={}", path.display());
-        println!("cargo:rustc-link-lib=static={lib}");
+        println!("cargo:rerun-if-changed={}", lib.display());
+        println!("cargo:rustc-link-lib=static={}", lib_name);
```
Comment (hasali19) [inline crates/dawn/build.rs:56]:
```suggestion
        println!("cargo:rustc-link-lib=static={lib_name}");
```
Simple variables can go directly inside the braces, which is a bit neater.

File: crates/dawn/src/lib.cpp [hunk 2]
```diff
@@ -6,33 +6,38 @@
...
+    auto instance = new dawn::native::Instance;
 
     // This makes things slow
     // instance->EnableBackendValidation(true);
-    // instance->SetBackendValidationLevel(dawn_native::BackendValidationLevel::Full);
+    // instance->SetBackendValidationLevel(dawn::native::BackendValidationLevel::Full);
 
-    instance->DiscoverDefaultAdapters();
+    /* DiscoverDefaultAdapters() was replaced with DiscoverPhysicalDevices()
+       which was in turn deprecated. TODO: check replacement code */
```
Comment (hasali19) [inline crates/dawn/src/lib.cpp:20]: Does this TODO still need to be addressed? If not, remove it.

File: crates/dawn/src/lib.cpp [hunk 3]
```diff
@@ -6,33 +6,38 @@
...
+    //instance->DiscoverDefaultAdapters();
```
Comment (hasali19) [inline crates/dawn/src/lib.cpp:21]: Any reason to keep the commented code?

File: crates/tint/build.rs [hunk 4]
```diff
@@ -23,23 +24,24 @@ fn main() -> Result<(), Box<dyn Error>> {
 
     println!("cargo:rustc-link-search=native={}", dawn_lib_dir.display());
 
-    let libs = ["tint_diagnostic_utils", "tint"];
-
+    let libs = fs::read_dir(dawn_lib_dir)
+        .unwrap()
```
Comment (hasali19) [inline crates/tint/build.rs:28]:
```suggestion
    let libs = fs::read_dir(dawn_lib_dir)?
```

File: crates/tint/build.rs [hunk 5]
```diff
@@ -23,23 +24,24 @@ fn main() -> Result<(), Box<dyn Error>> {
...
-        } else if target_family == "unix" {
-            format!("lib{lib}.a")
-        } else {
-            panic!("unsupported target_family '{target_family}'");
-        };
-
-        let path = dawn_lib_dir.join(lib_name);
+        
+        let lib_name = lib.file_stem().unwrap().to_str().unwrap();
+        let lib_name = &lib_name[3..];
```
Comment (hasali19) [inline crates/tint/build.rs:40]: Same comment about platform compatibility.

File: crates/tint/src/lib.cpp [hunk 6]
```diff
@@ -1,82 +1,91 @@
...
+    transform_inputs.Add<tint::ast::transform::Renamer::Config>(
+        tint::ast::transform::Renamer::Target::kMslKeywords,
         /* preserve_unicode */ false
     );
-    transform_manager.Add<tint::transform::Renamer>();
+    transform_manager.Add<tint::ast::transform::Renamer>();
 
-    auto transformed = transform_manager.Run(program.get(), std::move(transform_inputs));
-    if (!transformed.program.IsValid()) {
+    //auto transformed = transform_manager.Run(program.get(), std::move(transform_inputs));
```
Comment (hasali19) [inline crates/tint/src/lib.cpp:67]: Any reason to keep the commented code?

**Review by hasali19 — CHANGES_REQUESTED** [review]
Good stuff! Added a few mostly style related comments.

**Comment by hasali19** [comment]
Took the liberty of merging in some changes from `main` to enable the Actions workflow to run.

File: .cargo/config.toml
[the remark this answers is not in the document]
Comment (ambergorzynski) [reply .cargo/config.toml:1]: I thought this config file was necessary to pass the Dawn path but in fact it is not! Removed.

File: crates/ast/src/writer.rs [hunk 7]
```diff
@@ -98,12 +98,7 @@ impl Writer {
         for attr in &func.attrs {
             match attr {
                 FnAttr::Stage(stage) => {
-                    // TODO: Tint doesn't currently support the new stage attribute syntax - update when implemented
-                    if self.options.concise_stage_attrs {
```
[the remark this answers is not in the document]
Comment (ambergorzynski) [reply crates/ast/src/writer.rs:102]: Removed throughout

File: crates/dawn/build.rs [hunk 8]
```diff
@@ -24,56 +25,35 @@ fn main() -> Result<(), Box<dyn Error>> {
...
-        } else if target_family == "unix" {
-            format!("lib{lib}.a")
-        } else {
-            panic!("unsupported target_family '{target_family}'");
-        };
-
-        let path = dawn_lib_dir.join(lib_name);
+   
+        let lib_name = lib.file_stem().unwrap().to_str().unwrap();
+        let lib_name = &lib_name[3..];
```
[the remark this answers is not in the document]
Comment (ambergorzynski) [reply crates/dawn/build.rs]: Oops! No - removing the target-based naming was just a mistake. I have reinstated the previous target-based lib name. I have not tested it on a windows machine though.

File: crates/tint/build.rs (hunk 5 above)
Comment (ambergorzynski) [reply crates/tint/build.rs]: Changed as for the other case

**Comment by hasali19** [comment]
> Took the liberty of merging in some changes from `main` to enable the Actions workflow to run. Once you've updated the submodule, I'll mirror the change at https://github.com/wgslsmith/dawn-build, which will enable that version to be used in the workflow.

@ambergorzynski I've published a new dawn build. Could you update the version in `.github/dependency_versions.yaml` to `autobuild-2024-01-04-21-29`, which should enable the build and tests to pass. The checks are also failing [here](https://github.com/wgslsmith/wgslsmith/actions/runs/7411040004/job/20164661678?pr=27) due to some formatting issues I think, can you fix it please.

**Comment by ambergorzynski** [comment]
MacOS test is failing due to an undefined symbol "___isPlatformVersionAtLeast" in the Dawn source code. This looks like an issue with Xcode versions (e.g. see [here](https://forums.developer.apple.com/forums/thread/123003)). I can try specifying the Xcode version in ci.yml (although it looks like the default version in macos-latest should be sufficiently high) - does that sound OK or do you have other suggestions?

**Comment by hasali19** [comment]
Hmm, could also be an issue with how I'm building dawn for macos. Rather painful to debug since I don't actually have a mac. Give it a go, and if it doesn't work I'm happy to just disable the macos build and merge (assuming you don't need macos support).

**Comment by ambergorzynski** [comment]
OK, I'll try a few things and then give up if it's too tricky. I do have a mac, but I'm mainly using a Linux machine so no problem if we can't figure out mac support for now.

**Comment by ambergorzynski** [comment]
I tried building on my Mac and it also fails there, although with a slightly different error (cannot locate tint/dawn header files). I can spend some more time separately debugging this to figure out what the issues is, although I may not be super speedy.

**Comment by hasali19** [comment]
Sure, no problem. In the meantime do you think this is good to merge now if I disable the macos build?

**Comment by ambergorzynski** [comment]
Yes I think it's good to merge without the macOS build, thanks!

**Review by hasali19 — APPROVED** [review]

**Pull request merged**